### PR TITLE
standard erc20 trusttoken

### PR DESCRIPTION
**Pull request by hal909**

### CODE REVIEW

File: contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @dev This contract combines an upgradeability proxy with 
+ * basic authorization control functionalities
+ *
+ * This contract allows us to specify a time at which the proxy can no longer
+ * be upgraded
+ */
+contract TimeOwnedUpgradeabilityProxy is OwnedUpgradeabilityProxy {
+
+    uint256 expiration;
+    uint256 constant MAX_UINT = 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff;
```
Comment (krzysztof-jelski) [inline contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:16]:
```suggestion
    uint256 constant MAX_UINT = uint(-1);
```

File: contracts/TrustToken.sol [hunk 2]
```diff
@@ -11,21 +9,42 @@ import "./ValTokenWithHook.sol";
...
  */
-contract TrustToken is ValTokenWithHook, ClaimableContract {
+contract TrustToken is ERC20 {
+
+    /**
+     * @dev initalize trusttoken and give ownership to sender
+     * This is nessesary to set ownership for proxy 
+     */
+    function initalize() public {
+        owner_ = msg.sender;
```
Comment (krzysztof-jelski) [inline contracts/TrustToken.sol:19]: Needs flag to ensure it's run only once when deploying behind a proxy.

File: contracts/TrustToken.sol [hunk 3]
```diff
@@ -11,21 +9,42 @@ import "./ValTokenWithHook.sol";
...
+    }
+
+    /**
+     * @dev mint TRU
+     * Can never mint more than MAX_SUPPLY = 1.45 billion
+     */
+    function mint(address _to, uint256 _amount) external onlyOwner {
+        if (_totalSupply <= MAX_SUPPLY) {
```
Comment (krzysztof-jelski) [inline contracts/TrustToken.sol:27]:
```suggestion
        if (_totalSupply + _amount <= MAX_SUPPLY) {
```

File: contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+    * @dev the constructor sets the original owner of the contract to the sender account.
+    */
+    constructor() public {
+        _setUpgradeabilityOwner(msg.sender);
```
Comment (krzysztof-jelski) [inline contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:22]:
```suggestion
       super();
```

File: contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        expiration = MAX_UINT;
+    }
+
+    /**
+     *
+    */
+    function setExpiration(uint256 _newTimestamp) external onlyProxyOwner {
```
Comment (krzysztof-jelski) [inline contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:29]: Idea: perhaps from compliance/legal perspective it would be better to initialise expiration to in constructor (eg. +3months) and instead of `setExpiration` have `prolongExpiration`, which would change the `expiration` by a fixed value (eg. 1 month). The drawback of this solution is that we cannot "turn off" upgradability, which we can currently by setting `expiration` to a timestamp in the past. Anyway, it's your call @RafaelCosman to leave it as it is now or change it.

File: contracts/ERC20/ClaimableERC20.sol [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    */
+    function transferOwnership(address newOwner) public onlyOwner {
+        pendingOwner_ = newOwner;
+    }
+
+    /**
+    * @dev Allows the pendingOwner address to finalize the transfer.
+    */
+    function claimOwnership() public onlyPendingOwner {
+        address _pendingOwner = pendingOwner_;
```
Comment (krzysztof-jelski) [inline contracts/ERC20/ClaimableERC20.sol:63]: This local variable seems not necessary.

File: contracts/ERC20/ClaimableERC20.sol [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    modifier onlyOwner() {
+        require(msg.sender == owner_, "only owner");
+        _;
+    }
+
+    /**
+    * @dev Modifier throws if called by any account other than the pendingOwner.
+    */
+    modifier onlyPendingOwner() {
+        require(msg.sender == pendingOwner_);
```
Comment (krzysztof-jelski) [inline contracts/ERC20/ClaimableERC20.sol:47]:
```suggestion
        require(msg.sender == pendingOwner_, "only pending owner");
```

File: contracts/ERC20/ClaimableERC20.sol [hunk 8]
```diff
@@ -0,0 +1,68 @@
+pragma solidity ^0.5.13;
+
+import { ERC20ProxyStorage } from "./ERC20ProxyStorage.sol";
+
+/**
+ * @title ClaimableContract
+ * @dev The ClaimableContract contract is a copy of Claimable Contract by Zeppelin.
```
Comment (krzysztof-jelski) [inline contracts/ERC20/ClaimableERC20.sol:7]:
```suggestion
 * @dev The ClaimableContract contract is a copy of Claimable Contract by OpenZeppelin. 
```

File: contracts/TrustTokenProxyStorage.sol [hunk 9]
```diff
@@ -0,0 +1,38 @@
+pragma solidity ^0.5.13;
+
+import "@trusttoken/registry/contracts/Registry.sol";
+
+/**
+ * All storage must be declared here
+ * New storage must be appended to the end
+ * Never remove items from this list
+ */
+contract ProxyStorage {
```
Comment (krzysztof-jelski) [inline contracts/TrustTokenProxyStorage.sol:10]: Looks like this contract is not used anywhere.

File: contracts/ERC20/ERC20ProxyStorage.sol [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    uint256 constant MAX_SUPPLY = 145000000000000000;
+
+    mapping (address => uint256) _balances;
+    mapping (address => mapping (address => uint256)) _allowances;
+    uint256 _totalSupply;
+
+    address owner_;
+    address pendingOwner_;
+
+    /* Additionally, we have several keccak-based storage locations.
```
Comment (krzysztof-jelski) [inline contracts/ERC20/ERC20ProxyStorage.sol:19]: Are we actually using keccak-based storage in this repository? If not, we can safely remove this comment.

File: contracts/ERC20/StandardERC20.sol [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+     * total supply.
+     *
+     * Emits a {Transfer} event with `to` set to the zero address.
+     *
+     * Requirements
+     *
+     * - `account` cannot be the zero address.
+     * - `account` must have at least `amount` tokens.
+     */
+    function _burn(address account, uint256 amount) internal {
```
Comment (krzysztof-jelski) [inline contracts/ERC20/StandardERC20.sol:188]: I believe we can remove remove all the `_burn` functions, as they are not used. If we plan to utilise burning later, we can add with an upgrade.

Comment (RafaelCosman) [inline contracts/ERC20/StandardERC20.sol:188]: @hal909 who uses _burn? Do we need it?

File: contracts/ERC20/StandardERC20.sol [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+        emit Approval(owner, spender, amount);
+    }
+
+    /**
+     * @dev Destroys `amount` tokens from `account`.`amount` is then deducted
+     * from the caller's allowance.
+     *
+     * See {_burn} and {_approve}.
+     */
+    function _burnFrom(address account, uint256 amount) internal {
```
Comment (RafaelCosman) [inline contracts/ERC20/StandardERC20.sol:223]: @hal909 who uses _burnFrom? Do we need it?

File: contracts/Proxy/OwnedUpgradeabilityProxy.sol [hunk 13]
```diff
@@ -112,7 +112,7 @@ contract OwnedUpgradeabilityProxy {
     * @dev Allows the proxy owner to upgrade the current version of the proxy.
     * @param implementation representing the address of the new implementation to be set.
     */
-    function upgradeTo(address implementation) external onlyProxyOwner {
+    function upgradeTo(address implementation) public onlyProxyOwner {
```
Comment (RafaelCosman) [inline contracts/Proxy/OwnedUpgradeabilityProxy.sol:115]: @hal909 why did you change external to public here?

File: contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol (hunk 1 above)
Comment (RafaelCosman) [reply contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:16]: 👍

File: contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+    uint256 expiration;
+    uint256 constant MAX_UINT = 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff;
+
+    /**
+    * @dev the constructor sets the original owner of the contract to the sender account.
+    */
+    constructor() public {
+        _setUpgradeabilityOwner(msg.sender);
+        expiration = MAX_UINT;
```
Comment (RafaelCosman) [inline contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:23]: Can we have this start at 4 months rather than MAX_UINT and set later? I'd like that to be what people see when they look at the code, not MAX_UINT

File: contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /**
+     *
+    */
+    function setExpiration(uint256 _newTimestamp) external onlyProxyOwner {
+        expiration = _newTimestamp;
```
Comment (RafaelCosman) [inline contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:30]: This should also have and if() else or require(block.timestamp < expiration)

Comment (RafaelCosman) [reply contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:30]: Because once you're past the expiration you shouldn't be able to change the expiration

File: contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /**
+    * @dev Allows the proxy owner to upgrade the current version of the proxy.
+    * @param implementation representing the address of the new implementation to be set.
+    */
+    function upgradeTo(address implementation) public onlyProxyOwner {
+        if (block.timestamp < expiration) {
```
Comment (RafaelCosman) [inline contracts/Proxy/TimeOwnedUpgradeabilityProxy.sol:38]: Why use if else and not require()?

File: contracts/TrustToken.sol (hunk 3 above)
Comment (RafaelCosman) [inline contracts/TrustToken.sol:27]: @hal909 shouldn't this be _totalSupply+_amount <= MAX_SUPPLY?

Because otherwise you could exceed max amount. Also why if else and not require? Just curious, not sure that one or the other is better.

File: contracts/ERC20/ERC20ProxyStorage.sol [hunk 17]
```diff
@@ -0,0 +1,35 @@
+pragma solidity ^0.5.13;
+
+/**
+ * All storage must be declared here
+ * New storage must be appended to the end
+ * Never remove items from this list
+ */
+contract ERC20ProxyStorage {
+
+    uint256 constant MAX_SUPPLY = 145000000000000000;
```
Comment (RafaelCosman) [inline contracts/ERC20/ERC20ProxyStorage.sol:10]: @hal909 why do we need this when we also have the same thing in the other proxy storage contract?

**Comment by RafaelCosman** [comment]
I explicitly want to be able to reduce the expiration as well. So let's stay with set() instead of extend.

rafaelcosman@trusttoken.com
Co-Founder
TrustToken ( https://www.trusttoken.com/ ) - We're hiring! ( https://www.trusttoken.com/careers )

This message is intended only for the designated recipient(s).  It may contain confidential, privileged or proprietary information.  If you are not the designated recipient, you may not review, copy or distribute this message.  If you receive this message in error, please notify the sender by reply e-mail and delete this message.  Thank you.

**Comment by krzysztof-jelski** [comment]
Changes already included in other PRs.
